### ISSUE TITLE
Shell and python scripts are not running as expected over asyncssh

### ISSUE THREAD

**Issue by anarchy89** [ronf/asyncssh#554]
I am running the following script from my computer.

I have around 30 servers in the cloud that I am connecting to.

For each server I run a bash script which in turn runs a bunch of python scripts.

```
async def run_client(host, commands, **kwargs):
    output = ''

    async with asyncssh.connect(host, **kwargs) as conn:
        for i, command in enumerate(commands):
            result = await conn.run(command)

            if result.exit_status != 0:
                output += f'=== Command {i+1}: {command} failed ===\n' + result.stderr
            else:
                output += f'=== Command {i+1}: {command} succeeded ===\n' + result.stdout

        return output

async def run_multiple_clients():
    # Put your lists of hosts and commands here
    hosts = instance_ips
    commands = [f'python3 /home/ubuntu/Documents/reset.py', 'sleep 5', f'/home/ubuntu/Documents/misc/run.sh']

    tasks = (run_client(host, commands, username='ubuntu', client_keys=[SSH_KEY_PATH], known_hosts=None) for host in hosts)
    results = await asyncio.gather(*tasks, return_exceptions=True)

    for i, result in enumerate(results):
        if isinstance(result, Exception):
            print(f'Connection {i+1} to host {hosts[i]} failed: ' + str(result))
        else:
            print(f'Connection {i+1} to host {hosts[i]}:\n' + result, end='')

        print(75*'-')

asyncio.run(run_multiple_clients())
```

If I ssh into each server, and run `/home/ubuntu/Documents/misc/run.sh`,

Which is this,

```
#!/bin/bash
username='ubuntu'
counter=0

counter=$((counter + 1))
screen -dmS "f${counter}" sh -c "python3 /home/${username}/Documents/function1.py; exec bash"

counter=$((counter + 1))
screen -dmS "f${counter}" sh -c "python3 /home/${username}/Documents/function2.py; exec bash"

```

I am able to do `screen -R f1` and see the scripts running.

However,

IF i use the asyncssh script to run the shell script, the first part, the reset script runs properly, then, the screen instances are created but they don't run any command within it.

I even tried,

```
async def run_command(host, command, **kwargs):
    async with asyncssh.connect(host, **kwargs) as conn:
        result = await conn.run(command)
        return f'=== {host}: {result.stdout} ==='

async def run_commands_on_multiple_hosts():
    hosts = instance_ips
    command = f'/home/ubuntu/Documents/run.sh'

    tasks = [run_command(host, command, username='ubuntu', client_keys=[SSH_KEY_PATH], known_hosts=None) for host in hosts]
    results = await asyncio.gather(*tasks, return_exceptions=True)

    for i, result in enumerate(results):
        if isinstance(result, Exception):
            print(f'Error on host {hosts[i]}: {result}')
        else:
            print(result)

asyncio.run(run_commands_on_multiple_hosts())
```

Which should only run 1 command per host, I can see the print that each screen is being created, but nothing is being run in them.

The output even shows the command succeeded.

I don't understand why it's behaving this way, any help or insight on this would be greatly appreciated.

**Comment by ronf** [comment]
I've never tried using screen the way you are here, but in a quick test running on a local machine with no ssh involved at all, I can't quite make the "screen -dmS" work. It just exits immediately with no message and without appear to create a detached screen session. Does this work for you when you run screen by hand on a system without SSH?

**Comment by ronf** [comment]
Ok - I figured out that I was missing the "sh -c" in my test. With that, it worked better, but I couldn't attach using just the session label I assigned. I had to give "screen" an argument of the form <pid>.<label> to attach. Also, it was better to use "-r" there instead of "-R". If you use "-R" and get the session name wrong, it just starts a new session. Since we don't necessarily know the pid on the remote system, that could be a bit tricky. I'm not sure offhand if there's a way to tell screen not to include the pid in the session name.

I also suspect that quoting is going to be tricky when running a command like this through SSH, but I'll experiment with that a bit more and get back to you.

**Comment by ronf** [comment]
Ok - I got this to work here using both OpenSSH and AsyncSSH. With OpenSSH, the entire command needed to be in single quotes for things to work, like:

```
ssh localhost 'screen -dmS 1 sh -c "ls; exec bash"'
```

With AsyncSSH, it already did the right thing with the command as a string, so just the regular Python single quotes (since the command contains double quotes) worked fine:

```python
    async with asyncssh.connect('localhost') as conn:
        await conn.run('screen -dmS 1 sh -c "ls; exec bash"')
```

This created a detached session which I could see in 'screen -list', and from there I could get the full session name to pass into 'screen -r'.

**Comment by ronf** [comment]
Ah - I think the issue with the pid showing up in my tests may have been related to the fact that I already had a different screen session running on that host. So, you might not run into that if this is your only screen session.

**Comment by anarchy89** [comment]
Could you get it to work if you ran a shell script in the remote server which ran the screen ?

**Comment by ronf** [comment]
Yes - that works too:

```python
    async with asyncssh.connect('somehost') as conn:
        await conn.run('/tmp/run.sh')
```

where /tmp/run.sh is:

```
#!/bin/bash
counter=0

counter=$((counter + 1))
screen -dmS "f${counter}" sh -c "ls; exec bash"
```

Also, /tmp/run.sh needs to be marked as executable (chmod +x /tmp/run.sh).

Interestingly, it doesn't seem to work right when requesting a PTY. I'm not quite sure why. However, the default when providing a command to run and no term_type is to not request a PTY, and that case works. This is not an AsyncSSH issue, though. I saw the same thing with OpenSSH when running ssh with the '-t' option to request a PTY.

Perhaps something is going wrong in your function1.py and function2.py scripts? If you replace those with "ls", does it work?

**Comment by anarchy89** [comment]
Yeah, I have already marked all of them as executables before they were uploaded.

I'll give it a try with ls, and get back to you shortly. I am really confused I was on this for 12 hours yesterday trying to find out why it was just giving me blanks. Do you think it has something to do with the username='ubuntu', I am using?

**Comment by anarchy89** [comment]
I run echo test in each and it works. It seems it's just these python scripts that don't work, any ideas why?

If I run a python script out right it works. 

But if I run a shell script which runs a screen + python script it doesnt work. Could you give it a try? A python script that does a print('test') or something?

**Comment by anarchy89** [comment]
I have made progress! it seems my python scripts cannot access another python file causing the crash!!

Im not sure if you can figure out why.

So my directory looks like this.

variables
  |-- variables.py
function1
  |--  function1.py

variables.py looks like this

```
NUMBER_OF_X = 42
```

in function1.py it's

```
import sys
sys.path.insert(0, '../variables')

from variables import NUMBER_OF_X 

print(NUMBER_OF_X)
```
This script works fine if i run the shell script locally.

But when I run it from asyncssh, it give me an import error on variables. Any ideas on this?

**Comment by anarchy89** [comment]
I found a fix, i needed to import the variable differently and it doesn't crash anymore. thanks for talking me through it and troubleshooting with me.

I have another question,

```
async def run_client(host, commands, **kwargs):
    output = ''

    async with asyncssh.connect(host, **kwargs) as conn:
        for i, command in enumerate(commands):
            result = await conn.run(command)

            if result.exit_status != 0:
                output += f'=== Command {i+1}: {command} failed ===\n' + result.stderr
            else:
                output += f'=== Command {i+1}: {command} succeeded ===\n' + result.stdout

        return output

async def run_multiple_clients():
    # Put your lists of hosts and commands here
    hosts = instance_ips
    commands = [f'python3 /home/ubuntu/Documents/reset.py', 'sleep 5', f'/home/ubuntu/Documents/misc/run.sh']

    tasks = (run_client(host, commands, username='ubuntu', client_keys=[SSH_KEY_PATH], known_hosts=None) for host in hosts)
    results = await asyncio.gather(*tasks, return_exceptions=True)

    for i, result in enumerate(results):
        if isinstance(result, Exception):
            print(f'Connection {i+1} to host {hosts[i]} failed: ' + str(result))
        else:
            print(f'Connection {i+1} to host {hosts[i]}:\n' + result, end='')

        print(75*'-')

asyncio.run(run_multiple_clients())
```
this function,

can i confirm it into a function so that i can just insert the hosts and command everytime in a one liner like this.

```
async def run_client(host, commands, **kwargs):
    output = ''

    async with asyncssh.connect(host, **kwargs) as conn:
        for i, command in enumerate(commands):
            result = await conn.run(command)

            if result.exit_status != 0:
                output += f'=== Command {i+1}: {command} failed ===\n' + result.stderr
            else:
                output += f'=== Command {i+1}: {command} succeeded ===\n' + result.stdout

        return output

async def run_multiple_clients(hosts, commands):
    # Put your lists of hosts and commands here
    tasks = (run_client(host, commands, username='ubuntu', client_keys=[SSH_KEY_PATH], known_hosts=None) for host in hosts)
    results = await asyncio.gather(*tasks, return_exceptions=True)

    for i, result in enumerate(results):
        if isinstance(result, Exception):
            print(f'Connection {i+1} to host {hosts[i]} failed: ' + str(result))
        else:
            print(f'Connection {i+1} to host {hosts[i]}:\n' + result, end='')

        print(75*'-')

asyncio.run(run_multiple_clients(hosts, commands))
```

I tried it but I keep getting some asyncio errors.

**Comment by anarchy89** [comment]
Can I check something, the commands above, commands = [f'python3 /home/ubuntu/Documents/reset.py', 'sleep 5', f'/home/ubuntu/Documents/misc/run.sh'], are they run all at the same time? or are they run in sequence?

**Comment by anarchy89** [comment]
Now, half the servers connect and the other half give me this error, 

Connection 15 to host xx.xx.xx.xx failed: [Errno 61] Connect call failed ('xx.xx.xx.xx', 22

Any idea why I am getting this error?

**Comment by ronf** [comment]
Regarding use of "..", keep in mind that when you SSH into the remote server, any file references in the commands you send will be for paths on that server, and you would generally start in the remote user's home directory. So, a reference to ".." would be relative to that home directory, not the directing you are in on the client machine.

Regarding the commands, they should be run in sequence, waiting for one to finish before running the next. However, if you have multiple hosts defined, connections to the hosts should happen in parallel and each host will run the set of commands independently of one another. Different hosts could potentially be running different commands, as the sequencing only applies within each host.

The connect error you are seeing generally indicates that no SSH server is running on the target host (at least not on port 22), or it could indicate some kind of firewall in the way. Are you sure you have the right IPs? Is the set of failures consistent from one run to the next?

**Comment by anarchy89** [comment]
I figured it out, I needed to add a sleep because the servers were not fully turned on. It works as expected now.

So to confirm about multiple servers multiple commands. Within each server the commands are run in sequence (command1, waits for command 1 to finish, command 2, waits for command 2 to finish, command3, waits for command3 to finish), but among servers it's parallel/asynch correct ?

**Comment by ronf** [comment]
Yes, that's right.

If you look at the code, the "tasks" array in run_multiple_clients() contains a collection of coroutine calls to run_client(), with one entry per host. Those tasks are started and run in parallel due to the call to asyncio.gather(). That waits for all those coroutines to finish, but they're all run simultaneously.

Within run_client(), you can see there's a "for" loop that calls conn.run() on each command sequentially, and conn.run() waits for each command to finish before returning and allowing the next iteration of the "for" loop to run the next command.

**Comment by anarchy89** [comment]
Cool thanks for the help! Managed to fix everything.

**Comment by anarchy89** [comment]
@ronf Sometimes when I connect, I get a connection error if I send out the commands to fast, I think it's because the servers have not fully turned on. 

So to be safe I have added a time.sleep(60) to my script. Is there something that you have created in your code or some method I can use so that if I get a connection error I can wait until such time where it's safe and then send the commands ?

**Comment by ronf** [comment]
First thing: Don't use `time.sleep` from inside an async function. It will block all async tasks from running for the duration of the sleep. If you want to sleep from within an async task, using `await asyncio.sleep(delay)`.

There's no way to know whether a remote system is up before you try to connect to it, or to know how long before it will come up (if ever). So, the best you can do is to do a try...except to catch the connection error and then sleep for some time before trying again, using the async sleep mentioned above. You also probably want to count the number of times you tried, and give up after a certain number of tries so the script doesn't loop forever.

**Comment by anarchy89** [comment]
what is the connection error code ? so I can put it into the except block.

**Comment by ronf** [comment]
As far as I can tell, Python doesn't document all the possible errors that could be raised by the networking code here. However, in my experience, catching OSError and asyncssh.Error is good enough. By using OSError, you catch not only everything under ConnectionError, but also TimeoutError and other things like socket.gaierror when name resolution fails.

**Comment by anarchy89** [comment]
ok so if I am ssh'ing into 30+ servers using the above code, where should I put the try and except block, if 1 of the servers didn't get the command so I can keep resending it, and not send it again to the rest?

```
async def run_client(host, commands, **kwargs):
    output = ''

    async with asyncssh.connect(host, **kwargs) as conn:
        for i, command in enumerate(commands):
            result = await conn.run(command)

            if result.exit_status != 0:
                output += f'=== Command {i+1}: {command} failed ===\n' + result.stderr
            else:
                output += f'=== Command {i+1}: {command} succeeded ===\n' + result.stdout

        return output


async def run_multiple_clients(hosts, commands, **kwargs):
    tasks = (run_client(host, commands, **kwargs) for host in hosts)
    results = await asyncio.gather(*tasks, return_exceptions=True)

    for i, result in enumerate(results):
        if isinstance(result, Exception):
            print(f'Connection {i+1} to host {hosts[i]} failed: ' + str(result))
        else:
            print(f'Connection {i+1} to host {hosts[i]}:\n' + result, end='')

        print(75*'-')


asyncio.run(run_multiple_clients(hosts=hosts, commands=commands, username=USERNAME', client_keys=[SSH_KEY_PATH], known_hosts=None))
```

Should it be under the tasks? Like this?

```
async def run_multiple_clients(hosts, commands, **kwargs):
    try:
        tasks = (run_client(host, commands, **kwargs) for host in hosts)
        results = await asyncio.gather(*tasks, return_exceptions=True)
    except:
        tasks = (run_client(host, commands, **kwargs) for host in hosts)
        results = await asyncio.gather(*tasks, return_exceptions=True)

    for i, result in enumerate(results):
        if isinstance(result, Exception):
            print(f'Connection {i+1} to host {hosts[i]} failed: ' + str(result))
        else:
            print(f'Connection {i+1} to host {hosts[i]}:\n' + result, end='')

        print(75*'-')
```

sorry im not very sure about how async code works yet.

**Comment by ronf** [comment]
The simplest thing here would probably be to modify run_client() to add a retry mechanism, such as:

```python
async def run_client(host, commands, **kwargs):
    output = ''

    for retries in range(5):
        try:
            conn = await asyncssh.connect(host, **kwargs)
            break
        except (OSError, asyncssh.Error):
            pass
    else:
        return f'Connect to {host} failed\n'

    async with conn:
        for i, command in enumerate(commands):
            result = await conn.run(command)

            if result.exit_status != 0:
                output += f'=== Command {i+1}: {command} failed ===\n' + result.stderr
            else:
                output += f'=== Command {i+1}: {command} succeeded ===\n' + result.stdout

        return output
```

This will retry on host connection failures up to 5 times before giving up. If it succeeds on any of those attempts, it will behave as it did before and run the list of commands you provide. Command failures are still handled as they were before. You may want to await on `asyncio.sleep` for some time between retries, which you could do by replacing "pass" in the "except" block.

**Comment by anarchy89** [comment]
thanks! ill give this a try

**Comment by anarchy89** [comment]
@ronf i added a sleep, 
```
async def run_client(host, commands, **kwargs):
    output = ''

    for retries in range(60):
        try:
            conn = await asyncssh.connect(host, **kwargs)
            break
        except (OSError, asyncssh.Error):
            asyncio.sleep(10)
    else:
        return f'Connect to {host} failed\n'

    async with conn:
        for i, command in enumerate(commands):
            result = await conn.run(command)

            if result.exit_status != 0:
                output += f'=== Command {i+1}: {command} failed ===\n' + result.stderr
            else:
                output += f'=== Command {i+1}: {command} succeeded ===\n' + result.stdout

        return output
```

but i got this error, 

RuntimeWarning: coroutine 'sleep' was never awaited
  asyncio.sleep(10)
RuntimeWarning: Enable tracemalloc to get the object allocation traceback

So I am guessing i need to just add an await right?

```
async def run_client(host, commands, **kwargs):
    output = ''

    for retries in range(60):
        try:
            conn = await asyncssh.connect(host, **kwargs)
            break
        except (OSError, asyncssh.Error):
            await asyncio.sleep(10)
    else:
        return f'Connect to {host} failed\n'

    async with conn:
        for i, command in enumerate(commands):
            result = await conn.run(command)

            if result.exit_status != 0:
                output += f'=== Command {i+1}: {command} failed ===\n' + result.stderr
            else:
                output += f'=== Command {i+1}: {command} succeeded ===\n' + result.stdout

        return output
```

**Comment by ronf** [comment]
Yes - that's right. The error you are seeing is generally an indication of forgetting to add an `await` in front of a coroutine call.
